### DISCUSSION TITLE
Add pyramid levels using ImageMagick

### DISCUSSION THREAD

**Discussion by suminwei** [ImageMagick/ImageMagick#7250] [Help]
### ImageMagick version

7.1.0

### Operating system, version and so on

Linux

### Description

Hello,

I am trying to using ImageMagick to add pyramid levels to an svs file. I wasn't sure what the exact command would look like, and was hoping to ask for your help with some parameters. I think all I can figure out is I need to use the ```convert``` command. But other than that, I'm not sure what parameters I need to specify? I'd really appreciate your help!

Lisa

**Comment by snibgo** [comment]
IM does not have a coder to read or write SVS files.

**Reply by suminwei** [reply]
If I convert svs file to tiff, may you advise how to use ```convert``` command? Thank you.

**Comment by snibgo** [comment]
`convert` is an old command. For IM v7, which you have, you should use `magick` instead.

To create a tiff:
```
magick in.png -define tiff:tile-geometry=64x64 -compress LZW ptif:out.tiff
```
This reads a PNG file, and writes a pyramidical tiled TIFF file, with LZW compression, tile size 64x64 pixels.

IM can read the TIFF file, and write a PNG, or whatever you want:
```
magick out.tiff out2.png
```

See also https://stackoverflow.com/questions/75365808/is-it-possible-to-create-svs-files-at-all and http://www.andrewjanowczyk.com/converting-an-existing-image-into-an-openslide-compatible-format/ and https://web.archive.org/web/20120420105738/http://www.aperio.com/documents/api/Aperio_Digital_Slides_and_Third-party_data_interchange.pdf .

**Reply by suminwei** [reply]
Thank you!
When I ran the following command:
```
magick Slide11.tif -compress lzw -define tiff:tile-geometry=512x512 -define ptif:pyramid=512x4 -define tiff:subfiletype=reducedimage Slide11_pyramidi4.tiff
```
I get this error:
```
magick: unable to write pixel cache '/tmp/magick-CE1D5zyDXIj8f76GguhAU7cTxFLaPBGu': No space left on device @ error/cache.c/WritePixelCachePixels/5854.
```
How can I modify what the cache directory is if it's an issue of no space in the default cache?

**Comment by snibgo** [comment]
The problem is "No space left on device". I'm not sure which device, but I guess whatever `/tmp` is on.

**Comment by suminwei** [comment]
Ok, I set MAGICK_TMPDIR to a different path: 
```
export MAGICK_TMPDIR=/n/scratch/users/l/liw165/ImageMagick_tmp
```
and I ran the command again. I got a different error this time - it aborted at some point when running the following:
```
magick Slide11.tif -compress lzw -define tiff:tile-geometry=512x512 -define ptif:pyramid=512x4 -define tiff:subfiletype=reducedimage Slide11_pyramidi4.tiff
```

Are there any other variables I need to change? Thank you.

**Comment by fmw42** [comment]
How big is your input image dimensions?  Do you have enough RAM to hold the input and pyramid output in memory?  Does your policy.xml file have permissions for RAM, width and height and tmp space for your process?

Does the command above work for a smaller input and/or fewer levels.
